### ISSUE TITLE
Support ujson if available

### ISSUE THREAD

**Issue by seantalts** [stan-dev/cmdstanpy#38]
#### Summary:
If ujson (drop-in replacement for stdlib's json module) is available, use that instead as its several orders of magnitude faster. I think this is a 4 line change:

```python
try:
    import ujson as json
except ImportError: 
    import json
```

That should get us within @maedoc's 5% of fit time quite easily.

==================
WAS: request for faster data transfer
Ideas:
* Unix sockets (or in general allowing users to pass in the file handle for whatever they want to use to do communication between CmdStanPy and CmdStan). This should be a pretty quick fix to the code, just allowing for an additional argument.
* Something faster to parse and create than JSON (I know this would also require changes in CmdStan). This post compares a few options: https://yuhui-lin.github.io/blog/2017/08/01/serialization

**Comment by bob-carpenter** [comment]
I couldn't understand from the first bullet point where the sockets would go.  When would a user want to pass in a socket?  What function gets an additional argument?  The communication with CmdStan is going to be file-based, right?   I've been trying to resist the common R pattern of supplying multiple arguments, only one of which should be turned on.  I'd rather overload a single argument so the doc's less confusing (that is, doesn't have things like "this argument should have value `None` if this other argument has a value other than `None`.").

For the second bullet, it's not concrete enough to implement, so should probably be kicked to the forums for discussion before being promoted to an issue.

**Comment by seantalts** [comment]
I don't think all of this project's maintainers are active on discourse, so just wanted to get a conversation going on the repo itself. That said, I did mean this for brainstorming, so it's possible there are no action items.

I did a little googling and maybe the easiest thing to avoid transferring to physical media is to use this library (or its techniques): https://pypi.org/project/memory-tempfile/ (Unix domain sockets would require changing to a socket API and changes to cmdstan too I think, my bad.) It seems to be a drop-in replacement for NamedTemporaryFile that creates it on ramdisk/tmpfs space.

I think "not concrete enough to implement" is a fuzzy zone - it's possible someone could read up on a good 3rd data encoding scheme and just make a PR to add it, right? This isn't a case where we have to make one perfect choice - we could support a variety of formats.

**Comment by bob-carpenter** [comment]
To avoid further hijacking this thread, I created a forum post for the meta-discussion:
* [When is a feature request issue ready?](https://discourse.mc-stan.org/t/when-is-a-feature-request-issue-ready/8885)

**Comment by maedoc** [comment]
This came up in another issue on project scope, and I had previously done a prototype: one can use named pipes/fifos for the input data (json or otherwise) and output csv, and have the serialization/deserialization between CmdStan & CmdStanPy run concurrently.  Where this is substantial (in my experience) is for parsing output CSV incrementally for gigabytes of samples instead of having CmdStan write a big CSV file and waiting a long time for the Python parser to read it.

Implementing a binary format for CmdStan would be practical step forward, e.g. [BSON](https://en.wikipedia.org/wiki/BSON).  If CmdStan executables could use a binary format, one could even load the executable as a shared library and pass off data in a zero copy fashion, but it seems like fast interop should be using PyStan anyhow.

**Comment by seantalts** [comment]
> one can use named pipes/fifos for the input data (json or otherwise) and output csv, and have the serialization/deserialization between CmdStan & CmdStanPy run concurrently.

Ah yeah, I was thinking of named pipes, not unix domain sockets as the thing that basically acts like a file but is in memory. My bad! That's cool that you were already on it :) It seems like CmdStanPy could internally use named pipes whenever they're available on the platform we're on without any configuration from the user.

Re: shared library - I don't think Python users will typically have their data stored in memory the way that CmdStan needs it, so it seems like we'd have to do a copy most times anyway, right? Once you're doing a copy I think most of the overhead is in serializing to strings in JSON and writing to disk - I bet even if we just used BSON or msgpack (which are supposed to be basically drop-in replacements for JSON) and named pipes we'd get rid of ~90% of the existing overhead and be far enough for a while.

**Comment by bob-carpenter** [comment]
Protocol buffers were the other thing that'd been discussed as potentially being used cross-platform.

**Comment by ahartikainen** [comment]
Is this really a problem?

How long does each step take?

**Comment by seantalts** [comment]
For a 100,000x100 matrix it takes something like 12 seconds to rdump it out to a file with pycmdstan, I haven't been able to test cmdstanpy yet as there's no pip package and it's a bit of a pain. FWIW I measured named pipes and it seemed to take just about as long on my Mac laptop, so maybe that's not helpful after all.

**Comment by maedoc** [comment]
>  I haven't been able to test cmdstanpy yet as there's no pip package

`pip install git+https://github.com/stan-dev/cmdstanpy` should do the trick until a release is ready

> I measured named pipes and it seemed to take just about as long on my Mac laptop, so maybe that's not helpful after all.

the serialization is going to take time for that size data, the advantage of a fifo is concurrent serialization/deserialization. Any numbers should probably use a minimal model to put numbers in a context, e.g. `model { big_matrix ~ normal(0.0, 1.0); }` takes time to evaluate and if I/O is 5% of an optimization run, it's probably not worth worrying about.

**Comment by seantalts** [comment]
Thanks, that's very useful. And yeah, 100% agreed with all of your points (though a model needs parameters). I guess it's up to me to run the benchmark, so it looks like just serializing it (with jsondump in cmdstanpy - notably the rdump method there takes half as long) is about 10% of this model with N = 1,000,000:
```stan
data {
  int<lower=0> N;
  vector[N] x;
}
parameters {
  real mu;
}
model {
  mu ~ normal(0, 10);
  x ~ normal(mu, 1);
}
```
So if we ballpark estimate deserialization is of similar speed, then 20% of time is spent doing input and we could shave off half that by using a named pipe and doing it in parallel, another half by switching back to rdump, and then we're in the 5% range where maybe we don't care anymore. 

Here's my script and the output:
```python
import cmdstanpy
cmdstanpy.CMDSTAN_PATH="~/scm/clean/cmdstan"
import time
import functools

def timed(f, key):
    @functools.wraps(f)
    def wrapper(*args, **kwargs):
        start = time.time()
        res = f(*args, **kwargs)
        end = time.time()
        print("{} took {}s".format(key, end-start))
        return res
    return wrapper

import cmdstanpy.utils
cmdstanpy.utils.jsondump = timed(cmdstanpy.utils.jsondump, "jsondump")
cmdstanpy.utils.rdump = timed(cmdstanpy.utils.rdump, "rdump")

from cmdstanpy.cmds import compile_model, sample
import numpy as np

def expit(x): return 1/(1+np.exp(-x))

sample = timed(sample, "sample")
model = compile_model("silly.stan")
N = 1000 * 1000 
x = np.random.normal(0, 1, size=N)
cmdstanpy.utils.rdump("silly.Rdata", dict(N=N, x=x))
runs = sample(model, 
        data=dict(N=N, x=x),
        chains=1)
```

```
halair ~/scm/tfp  $ python3.7 silly.py 
rdump took 1.103652000427246s
input data tempfile: /var/folders/zm/pplrl2mj38xg256cbjb_jjt40000gn/T/tmpe92hvpjk/tmp9mzu8f9g.json
jsondump took 2.062939167022705s
start chain 1.  
finish chain 1.  
sample took 22.398995876312256s
deleting tmpfiles dir: /var/folders/zm/pplrl2mj38xg256cbjb_jjt40000gn/T/tmpe92hvpjk
done
```

(and cmdstan reports 20s of sampling, suggesting that extra 2s is spent deserializing json).

**Comment by seantalts** [comment]
Might also be worth noting that jsondump scales particularly poorly with additional structure: splitting the 1M elements into 500k rows of 2 columns makes the jsondump go from 2s to 3.4s.

**Comment by bob-carpenter** [comment]
I did a baseline evaluation using clang++ at -O3 using `<<` to print out ASCII and `write` to output bytes.  For a 5000 x 5000 matrix  (25M entries) and 1000 x 1000 matrix (1M entries)

* `<<` performance: 1M at 0.85s, 25M at 20s
* `write` performance: 1M at 0.01s, 25M at 0.5s

So it's definitely Python getting in the way.  

This is really a best case version for binary.  This is not doing any kind of correction for platform representation and just dumps the raw bytes.  The `<<` performance is probably not ideal for converting to ASCII.

Here's the code I used to test:

```
#include <iomanip>
#include <iostream>
#include <fstream>
#include <vector>
#include <chrono>

int main() {
  using std::fstream;
  using std::vector;
  using std::chrono::high_resolution_clock;
  using std::chrono::microseconds;
  using std::chrono::duration_cast;
  using std::setprecision;

  int M = 1000;
  int N = 1000;
  vector<double> x(M * N);
  for (int i = 0; i < x.size(); ++i)
    x[i] = i * -2.37826587365472;
  auto start = high_resolution_clock::now();
  fstream fs("foo.txt", fstream::out);
  fs << setprecision(15);
  for (int j = 0; j < N; ++j)
    for (int i = 0; i < M; ++i) {
      fs << x[M * i + j] << ',';
    fs << '\n';
  }
  fs.close();
  auto end = high_resolution_clock::now();
  double seconds = duration_cast<microseconds>(end - start).count() / 1e6;
  std::cout << "ASCII: elapsed write time = "
            << seconds << " seconds" << std::endl;

  start = high_resolution_clock::now();
  fstream bs("bar.txt", fstream::out);
  char* xd = reinterpret_cast<char*>(x.data());
  bs.write(xd, x.size() * sizeof(double));
  bs.close();
  end = high_resolution_clock::now();
  seconds = duration_cast<microseconds>(end - start).count() / 1e6;
  std::cout << "raw memory: elapsed write time = "
            << seconds << " seconds" << std::endl;

}
```

and I ran it with:

```
$ clang++ -std=c++1y -O3 cio.cpp
$ ./a.out
```

**Comment by ahartikainen** [comment]
@seantalts have you tried `ujson`?

**Comment by seantalts** [comment]
Yeah, I think we weren't worried about raw file I/O anymore, just using named pipes so we can serialize and deserialize at the same time. Those operations are expensive in Python and C++ for formats like JSON - to bench that statement: 

cmdstan json: 22.7s reported by `time`, 18.95s reported by cmdstan
cmdstan rdump: 20.3s reported by `time`, 18.38s reported by cmdstan
(for comparison - pycmdstan json: 24.8s in `sample`)

(all with seed set to 1, assuming that is working).

So to summarize I think there is expensive serialization due to the formats we've chosen and we can alleviate that with named pipes (to do serialization/deserialization at the same time) or, and I would actually just recommend this, we could switch to a faster format like `msgpack`. On my machine
rdump - 1.1s
json-2.0s, and scales very poorly with structure
msgpack-0.04s

I haven't tried ujson but I suspect the json parser @mitzimorris wrote in C++ has pretty good performance and we're mostly limited by the codec / format of JSON here.

**Comment by ahartikainen** [comment]
I don't really worry about those numbers, but I think the bottleneck for json / rdump is found in Python side, not Stan side. That's why `ujson` could work a bit better.

That said, I personally think `msgpack` is the correct way to go. Given it can be compressed, helps with the large files.

Other fast formats that are used with pydata stack are `feather`, `parquet`, `netcdf4`, `hdf5`. But I like the msgpack given it works with nD arrays and is simple to install.

**Comment by seantalts** [comment]
`ujson` is much faster for this example:
```
msgpack.dump took 0.09357094764709473s
ujson.dump took 0.14149093627929688s
rdump took 1.1772470474243164s
jsondump took 2.1471049785614014s
```

**Comment by seantalts** [comment]
splitting it into two columns and msgpack actually does worse than ujson :P 
```
msgpack.dump took 0.36424708366394043s
ujson.dump took 0.20448803901672363s
rdump took 1.2897250652313232s
jsondump took 3.2851688861846924s
```

**Comment by seantalts** [comment]
So maybe the easiest thing to do here to get within that 5% range is to just switch to `ujson` if it's available. A pattern I've seen quite a few times back in the day was something like:
```python
try:
    import ujson as json
except ImportError: 
    import json
```

**Comment by ahartikainen** [comment]
Yes, that sounds reasonable.

**Comment by bob-carpenter** [comment]
> So to summarize I think there is expensive serialization due to the formats we've chosen 

I think @ahartikainen has already taken us here, but what I was trying to show is that the bottleneck was the inefficient implementation of the format, not the format itself.  ASCII is much slower than binary (factor of 100 or so at full precision), but I don't think the ASCII would be a bottleneck in most of our processing (the only place I've ever seen it is in fitting a huge linear regression with optimization---then I/O can dominate because the fit's so fast).

I do agree that it'd be nice to have a binary format at some point.  If nothing else, it'll save us about 50% in space, which isn't nothing, especially when running on things like clusters with relatively slow file I/O.

it's not the ASCII format that's the bottleneckso much as the implementation of the format that's the bottleneck.  The raw compute needed to decode is relatively small, though still about 100 times slower than binary if we print ASCII with 16 digits of precision.

**Comment by seantalts** [comment]
Okay, I'll change the issue to just ask that we support ujson for now at least if its available.
